### PR TITLE
SITES-319: Draft of flagging php in database

### PR DESCRIPTION
**Pull request by kbrownell**
# READY FOR REVIEW

# Summary
- We want to find and abort the migration of sites with PHP in the database.  If there is a standard way of parsing data for PHP, please let me know (this has proven a difficult think to search for as it returns parsing functions in PHP).  Based on a small subset of test databases, the following regex seemed to do a pretty good job of flagging PHP, while excluding non sequiturs.  But we will likely need to add more over time.

```
    - '.\{0,20\}$[^S$D|\.|0-9|conf|user|_SESSION][A-Za-z].\{0,20\}'
    - '.\{0,20\}<?[^xml].\{0,20\}'
    - 'eval('
    - 'system('
    - 'preg_replace('
    - 'create_function'
    - 'include\['
    - 'require\['
```

# Needed By (2/1)
- End of sprint

# Urgency
- While important, not urgent.

# Steps to Test

1. I've been testing with SITES224 and eao.  SITES224 should log the following lines:

```
        "lay_format\\\";s:46:\\\"$display_output = $enti",
        "lay_format\\\";s:46:\\\"$display_output = $enti",
        "lay_format\\\";s:46:\\\"$display_output = $enti",
        "lay_format\\\";s:46:\\\"$display_output = $enti"
```

2. Only the eao host should move forward with the migration.
3. Check that the PHP above is logged in /tmp/ansible.log.

# Affected Projects or Products
- Sites 2.0 security efforts

# Associated Issues and/or People
- https://stanfordits.atlassian.net/browse/SITES-319
- Thanks to Dmitry for helping me figure out how to log the grep results before I fail and abandon the host with PHP in the database.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by jbickar** [comment]
List of exploitable functions from Acquia:

```
exec(
passthru(
shell_exec(
system(
proc_open(
popen(
curl_exec(
curl_multi_exec(
parse_ini_file(
show_source(
```

**Comment by kbrownell** [comment]
Hello John, I added the callbacks Shea found at: https://cgit.drupalcode.org/site_audit/tree/Check/Security/MenuRouter.php?h=8.x-2.x#n92.  But then later excluded: file, link, and mail because there were a number instances that started to become tricky to exclude, where these words appeared in text.

Running the full set takes about 10m per site.  Let me know if we want to revisit the list to cut down that time.  Also, I know we had tried to double up include and include_once/require and require_once, but with the additional of the parens options appended to the end, I chose to separate them out instead of troubleshoot further to make them continue working within the same grep.

Running this on SITES-207F and eao, failed multiple times for SITES-207F (yay) and passed for eao (double yay).  The results can be found in /tmp/ansible.log after a run.

**Comment by kbrownell** [comment]
John asked we rewrite the default behavior to allow us to specify in the vars file or a run of the playbook, whether we'd prefer the playbook to fail, continue, or ask by host what to do if we come across what looks like PHP in the database.

**Comment by kbrownell** [comment]
@jbickar I added a new variable to default.migration_vars.yml with three options: fail, continue, or prompt for what action to take on hosts with what looks like PHP in their database.  To test this, I might recommend adding this new `php_candidates_consequence` variable to your migration_vars.yml file and trying:

1. continue, `SITES-207F` and `eao` should continue (assuming you're testing with these two sites).
2. fail, `SITES-207F` should abort and `eao` should continue.
3. prompt and then enter continue, `SITES-207F` and `eao` should continue.
4. prompt and then enter fail, `SITES-207F` should abort and `eao` should continue.

**Comment by jbickar** [comment]
With `php_candidates_consequence: "fail"`:

```
drupal7-update-status      : ok=32   changed=14   unreachable=0    failed=0   
jbickar                    : ok=7    changed=3    unreachable=0    failed=1   
nmardoin                   : ok=7    changed=3    unreachable=0    failed=1   
sheamck                    : ok=30   changed=14   unreachable=0    failed=0   
swsblog                    : ok=7    changed=5    unreachable=0    failed=1  
```

**Comment by jbickar** [comment]
With `php_candidates_consequence: "continue", I get errors:

```
TASK [check-php : Save PHP in file] **************************************************************************************************************************************************************************************************
skipping: [drupal7-update-status]
skipping: [sheamck]
fatal: [swsblog]: FAILED! => {"failed": true, "msg": "The conditional check 'php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"' failed. The error was: error while evaluating conditional (php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"): 'php_candidates_consequence_prompted' is undefined"}
fatal: [jbickar]: FAILED! => {"failed": true, "msg": "The conditional check 'php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"' failed. The error was: error while evaluating conditional (php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"): 'php_candidates_consequence_prompted' is undefined"}
fatal: [nmardoin]: FAILED! => {"failed": true, "msg": "The conditional check 'php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"' failed. The error was: error while evaluating conditional (php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"): 'php_candidates_consequence_prompted' is undefined"}
```

**Comment by jbickar** [comment]
Same with `php_candidates_consequence: "prompt"`:

```
TASK [check-php : Save PHP in file] **************************************************************************************************************************************************************************************************
skipping: [drupal7-update-status]
skipping: [sheamck]
fatal: [jbickar]: FAILED! => {"failed": true, "msg": "The conditional check 'php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"' failed. The error was: error while evaluating conditional (php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"): 'php_candidates_consequence_prompted' is undefined"}
fatal: [swsblog]: FAILED! => {"failed": true, "msg": "The conditional check 'php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"' failed. The error was: error while evaluating conditional (php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"): 'php_candidates_consequence_prompted' is undefined"}
fatal: [nmardoin]: FAILED! => {"failed": true, "msg": "The conditional check 'php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"' failed. The error was: error while evaluating conditional (php_candidates_consequence == \"fail\" or php_candidates_consequence_prompted.user_input == \"fail\"): 'php_candidates_consequence_prompted' is undefined"}
```

**Comment by jbickar** [comment]
I am now getting the error `The task includes an option with an undefined variable. The error was: 'php_candidates_consequence_prompted' is undefined`

**Comment by kbrownell** [comment]
Removed prompt option and tested that hosts failed that should fail and continued when the user requested they continue.  But of course, it's good to have someone else test too!

File: roles/check-php/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,180 @@
+---
+# [^ >=] is for system, but I found it only worked after the parens
+parens: '[ ]\{0,\}([^ >=]'
+php_syntax:
+  - '$[^S$D|\.|0-9|conf|user|_SESSION][A-Za-z]'
```
Comment (jbickar) [inline roles/check-php/defaults/main.yml]: Every Earth site is failing on this check, with results like this:

```
methane [[u'oP\x17$poj', u'N${G', u'$\x0es\x04\x1fEF\u05ca', u'oP\x17$pgj', u'\x1f3\u0421\x12$,t', u'5\x12cL$\x15e'], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], []]
```

Let me find where that is in the DB dump.

Comment (jbickar) [reply roles/check-php/defaults/main.yml]: Looks like it's in a cache table. So I think what we need to do is call `sql-dump` with `--structure-tables-key=common`. Will test that.

Comment (jbickar) [reply roles/check-php/defaults/main.yml]: c66b7a1 fixes that.

**Comment by kbrownell** [comment]
`sql-dump --structure-tables-key=common` worked just fine for me.  And `dekaslab` passed, as it should.

**Pull request merged**